src/Components/Server/src/Circuits/ArrayBuilderMemoryStream.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Web.Rendering
+{
+    /// <summary>
+    /// Writeable memory stream backed by a an <see cref="ArrayPool{T}"/>.
+    /// </summary>
+    internal sealed class ArrayBuilderMemoryStream : Stream
```

Comment:
Does this produce a better result than an `ObjectPool<MemoryStream>`?

Reply:
I'm guessing the difference is that `ObjectPool<MemoryStream>` wouldn't know ever shrink the underlying buffers or know to prioritize releasing the larger ones, so over time memory usage would tend to grow until all the pooled buffers were as big as the largest ever batches.

`ArrayBuilderMemoryStream` does better because the underlying `ArrayPool` is aware of the sizes of the buffers and knows to match the best-sized ones with whatever builder is requesting a buffer of that kind of size, right? And theoretically it could do smart things to release the buffers that were of infrequently used sizes, though no idea if it does.

Reply:
I hadn't thought about it. There's enough APIs on MemoryStream to pull this off without much ceremony. It'd be hard to tell which one's better for this scenario without doing some profiling. Unfortunately, I'm currently blocked on being able to do any more profiling. 

Do you think it's worthwhile following up on this or could we consider this good enough?

Reply:
There are reasons to think what you've already done here, backed by `ArrayPool` should be able to produce better results than an `ObjectPool<MemoryStream>` so I'd keep it as-is (lacking any other reason to change it).